core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,438 @@
+  private val shuttingDown = new AtomicBoolean(false)
+
+  /** lock protecting access to loading and owned partition sets */
+  private val partitionLock = new ReentrantLock()
+
+  /** partitions of transaction topic that are assigned to this manager, partition lock should be called BEFORE accessing this set */
+  private val ownedPartitions: mutable.Map[Int, Int] = mutable.Map()
+
+  /** partitions of transaction topic that are being loaded, partition lock should be called BEFORE accessing this set */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /** transaction metadata cache indexed by transactional id */
+  private val transactionMetadataCache = new Pool[String, TransactionMetadata]
+
+  /** number of partitions for the transaction log topic */
+  private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
+
+  def enablePidExpiration() {
+    scheduler.startup()
+
+    // TODO: add transaction and pid expiration logic
+  }
+
+  /**
+   * Get the transaction metadata associated with the given transactional id, or null if not found
+   */
+  def getTransaction(transactionalId: String): Option[TransactionMetadata] = {
+    Option(transactionMetadataCache.get(transactionalId))
+  }
+
+  /**
+   * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
+   */
+  def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): TransactionMetadata = {
+    val currentTxnMetadata = transactionMetadataCache.putIfNotExists(transactionalId, txnMetadata)
+    if (currentTxnMetadata != null) {
+      currentTxnMetadata
+    } else {
+      txnMetadata
+    }
+  }
+
+  /**
+   * Validate the given transaction timeout value
+   */
+  def validateTransactionTimeoutMs(txnTimeoutMs: Int): Boolean =
+    txnTimeoutMs <= config.transactionMaxTimeoutMs && txnTimeoutMs > 0
+
+  def transactionTopicConfigs: Properties = {
+    val props = new Properties
+
+    // enforce disabled unclean leader election, no compression types, and compact cleanup policy
+    props.put(LogConfig.UncleanLeaderElectionEnableProp, "false")
+    props.put(LogConfig.CompressionTypeProp, NoCompressionCodec)
+    props.put(LogConfig.CleanupPolicyProp, LogConfig.Compact)
+
+    props.put(LogConfig.MinInSyncReplicasProp, config.transactionLogMinInsyncReplicas.toString)
+    props.put(LogConfig.SegmentBytesProp, config.transactionLogSegmentBytes.toString)
+
+    props
+  }
+
+  def partitionFor(transactionalId: String): Int = Utils.abs(transactionalId.hashCode) % transactionTopicPartitionCount
+
+  def coordinatorEpochFor(transactionId: String): Option[Int] = inLock (partitionLock) {
+    ownedPartitions.get(partitionFor(transactionId))
+  }
+
+  def isCoordinatorFor(transactionalId: String): Boolean = inLock(partitionLock) {
+    val partitionId = partitionFor(transactionalId)
+
+    // partition id should be within the owned list and NOT in the corrupted list
```

Comment:
The comment is out of date. The original version of the patch had a notion of 'corrupted partitions', but the definition what is corrupted was not clear and we dropped the notion altogether.